// react-native-gesture-handler/lib/commonjs/web/handlers/FlingGestureHandler.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _State = require("../../State");

var _constants = require("../constants");

var _GestureHandler = _interopRequireDefault(require("./GestureHandler"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const DEFAULT_MAX_DURATION_MS = 800;
const DEFAULT_MIN_ACCEPTABLE_DELTA = 160;
const DEFAULT_DIRECTION = _constants.Direction.RIGHT;
const DEFAULT_NUMBER_OF_TOUCHES_REQUIRED = 1;

class FlingGestureHandler extends _GestureHandler.default {
  constructor(...args) {
    super(...args);

    _defineProperty(this, "numberOfPointersRequired", DEFAULT_NUMBER_OF_TOUCHES_REQUIRED);

    _defineProperty(this, "direction", DEFAULT_DIRECTION);

    _defineProperty(this, "maxDurationMs", DEFAULT_MAX_DURATION_MS);

    _defineProperty(this, "minAcceptableDelta", DEFAULT_MIN_ACCEPTABLE_DELTA);

    _defineProperty(this, "delayTimeout", void 0);

    _defineProperty(this, "startX", 0);

    _defineProperty(this, "startY", 0);

    _defineProperty(this, "maxNumberOfPointersSimultaneously", 0);
  }

  init(ref, propsRef) {
    super.init(ref, propsRef);
  }

  updateGestureConfig({
    enabled = true,
    ...props
  }) {
    super.updateGestureConfig({
      enabled: enabled,
      ...props
    });
    this.enabled = enabled;

    if (this.config.direction) {
      this.direction = this.config.direction;
    }

    if (this.config.numberOfPointers) {
      this.numberOfPointersRequired = this.config.numberOfPointers;
    }
  }

  transformNativeEvent(event) {
    return {
      x: event.offsetX,
      y: event.offsetY,
      absoluteX: event.x,
      absoluteY: event.y
    };
  }

  startFling(event) {
    this.startX = event.x;
    this.startY = event.y;
    this.begin(event);
    this.maxNumberOfPointersSimultaneously = 1;
    this.delayTimeout = setTimeout(() => this.fail(event), this.maxDurationMs);
  }

  tryEndFling(event) {
    if (this.maxNumberOfPointersSimultaneously === this.numberOfPointersRequired && (this.direction & _constants.Direction.RIGHT && event.x - this.startX > this.minAcceptableDelta || this.direction & _constants.Direction.LEFT && this.startX - event.x > this.minAcceptableDelta || this.direction & _constants.Direction.UP && this.startY - event.y > this.minAcceptableDelta || this.direction & _constants.Direction.DOWN && event.y - this.startY > this.minAcceptableDelta)) {
      clearTimeout(this.delayTimeout);
      this.activate(event);
      return true;
    }

    return false;
  }

  endFling(event) {
    if (!this.tryEndFling(event)) {
      this.fail(event);
    }
  }

  onPointerDown(event) {
    super.onPointerDown(event);
    this.tracker.addToTracker(event);

    if (this.currentState === _State.State.UNDETERMINED) {
      this.startFling(event);
    }

    if (this.currentState !== _State.State.BEGAN) {
      return;
    }

    this.tryEndFling(event);

    if (this.tracker.getTrackedPointersCount() > this.maxNumberOfPointersSimultaneously) {
      this.maxNumberOfPointersSimultaneously = this.tracker.getTrackedPointersCount();
    }
  }

  onPointerMove(event) {
    this.tracker.track(event);

    if (this.currentState !== _State.State.BEGAN) {
      return;
    }

    this.tryEndFling(event);
    super.onPointerMove(event);
  }

  onPointerUp(event) {
    this.tracker.removeFromTracker(event.pointerId);

    if (this.currentState !== _State.State.BEGAN) {
      return;
    }

    this.endFling(event);
  }

  activate(event, force) {
    super.activate(event, force);
    this.end(event);
  }

  resetConfig() {
    super.resetConfig();
    this.numberOfPointersRequired = DEFAULT_NUMBER_OF_TOUCHES_REQUIRED;
    this.direction = DEFAULT_DIRECTION;
  }

}

exports.default = FlingGestureHandler;
