// react-native-gesture-handler/lib/commonjs/web/handlers/RotationGestureHandler.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _State = require("../../State");

var _interfaces = require("../interfaces");

var _GestureHandler = _interopRequireDefault(require("./GestureHandler"));

var _RotationGestureDetector = _interopRequireDefault(require("../detectors/RotationGestureDetector"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const ROTATION_RECOGNITION_THRESHOLD = Math.PI / 36;

class RotationGestureHandler extends _GestureHandler.default {
  constructor(...args) {
    super(...args);

    _defineProperty(this, "rotation", 0);

    _defineProperty(this, "velocity", 0);

    _defineProperty(this, "cachedAnchorX", 0);

    _defineProperty(this, "cachedAnchorY", 0);

    _defineProperty(this, "rotationGestureListener", {
      onRotationBegin: _detector => true,
      onRotation: (detector, event) => {
        const previousRotation = this.rotation;
        this.rotation += detector.getRotation();
        const delta = detector.getTimeDelta();

        if (delta > 0) {
          this.velocity = (this.rotation - previousRotation) / delta;
        }

        if (Math.abs(this.rotation) >= ROTATION_RECOGNITION_THRESHOLD && this.currentState === _State.State.BEGAN) {
          this.activate(event);
        }

        return true;
      },
      onRotationEnd: (_detector, event) => {
        this.end(event);
      }
    });

    _defineProperty(this, "rotationGestureDetector", new _RotationGestureDetector.default(this.rotationGestureListener));
  }

  init(ref, propsRef) {
    super.init(ref, propsRef);
    this.setShouldCancelWhenOutside(false);
  }

  updateGestureConfig({
    enabled = true,
    ...props
  }) {
    super.updateGestureConfig({
      enabled: enabled,
      ...props
    });
    this.enabled = enabled;
  }

  transformNativeEvent(_event) {
    return {
      rotation: this.rotation ? this.rotation : 0,
      anchorX: this.getAnchorX(),
      anchorY: this.getAnchorY(),
      velocity: this.velocity ? this.velocity : 0
    };
  }

  getAnchorX() {
    const anchorX = this.rotationGestureDetector.getAnchorX();
    return anchorX ? anchorX : this.cachedAnchorX;
  }

  getAnchorY() {
    const anchorY = this.rotationGestureDetector.getAnchorY();
    return anchorY ? anchorY : this.cachedAnchorY;
  }

  onPointerDown(event) {
    super.onPointerDown(event);
    this.tracker.addToTracker(event);

    if (this.tracker.getTrackedPointersCount() <= 1) {
      return;
    }

    this.tryBegin(event);
    this.rotationGestureDetector.onTouchEvent(event, this.tracker);
  }

  onPointerMove(event) {
    if (this.tracker.getTrackedPointersCount() < 2 || !this.rotationGestureDetector) {
      return;
    }

    if (this.getAnchorX()) {
      this.cachedAnchorX = this.getAnchorX();
    }

    if (this.getAnchorY()) {
      this.cachedAnchorY = this.getAnchorY();
    }

    this.tracker.track(event);
    this.rotationGestureDetector.onTouchEvent(event, this.tracker);
    super.onPointerMove(event);
  }

  onPointerUp(event) {
    if (!this.rotationGestureDetector) {
      this.tracker.resetTracker();
      return;
    }

    if (this.tracker.getTrackedPointersCount() > 1) {
      this.rotationGestureDetector.onTouchEvent(event, this.tracker);
      this.tracker.removeFromTracker(event.pointerId);
    } else {
      this.tracker.removeFromTracker(event.pointerId);
      this.rotationGestureDetector.onTouchEvent(event, this.tracker);

      if (this.currentState !== _State.State.ACTIVE) {
        return;
      }
    }

    if (event.eventType !== _interfaces.EventTypes.UP) {
      return;
    }

    if (this.currentState === _State.State.ACTIVE) {
      this.end(event);
    } else {
      this.fail(event);
    }
  }

  onPointerCancel(event) {
    this.end(event);
    this.reset();
  }

  tryBegin(event) {
    if (this.currentState !== _State.State.UNDETERMINED) {
      return;
    }

    this.resetProgress();
    this.begin(event);
  }

  activate(event, _force) {
    if (this.currentState !== _State.State.ACTIVE) {
      this.resetProgress();
    }

    super.activate(event);
  }

  onReset() {
    if (this.currentState === _State.State.ACTIVE) {
      return;
    }

    this.rotation = 0;
    this.velocity = 0;
  }

}

exports.default = RotationGestureHandler;
