// react-native-gesture-handler/lib/commonjs/web/tools/InteractionManager.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class InteractionManager {
  constructor() {
    _defineProperty(this, "waitForRelations", new Map());

    _defineProperty(this, "simultaneousRelations", new Map());
  }

  configureInteractions(handler, config) {
    this.dropRelationsForHandlerWithTag(handler.getTag());

    if (config.waitFor) {
      const waitFor = [];
      config.waitFor.forEach(handler => {
        if (typeof handler === 'number') {
          waitFor.push(handler);
        } else {
          waitFor.push(handler.handlerTag);
        }
      });
      this.waitForRelations.set(handler.getTag(), waitFor);
    }

    if (config.simultaneousHandlers) {
      const simultaneousHandlers = [];
      config.simultaneousHandlers.forEach(handler => {
        if (typeof handler === 'number') {
          simultaneousHandlers.push(handler);
        } else {
          simultaneousHandlers.push(handler.handlerTag);
        }
      });
      this.simultaneousRelations.set(handler.getTag(), simultaneousHandlers);
    }

    handler.setInteractionManager(this);
  }

  shouldWaitForHandlerFailure(handler, otherHandler) {
    const waitFor = this.waitForRelations.get(handler.getTag());

    if (!waitFor) {
      return false;
    }

    let shouldWait = false;
    waitFor.forEach(tag => {
      if (tag === otherHandler.getTag()) {
        shouldWait = true;
        return; //Returns from callback
      }
    });
    return shouldWait;
  }

  shouldRecognizeSimultaneously(handler, otherHandler) {
    const simultaneousHandlers = this.simultaneousRelations.get(handler.getTag());

    if (!simultaneousHandlers) {
      return false;
    }

    let shouldRecognizeSimultaneously = false;
    simultaneousHandlers.forEach(tag => {
      if (tag === otherHandler.getTag()) {
        shouldRecognizeSimultaneously = true;
        return;
      }
    });
    return shouldRecognizeSimultaneously;
  }

  shouldRequireHandlerToWaitForFailure(_handler, _otherHandler) {
    return false;
  }

  shouldHandlerBeCancelledBy(_handler, _otherHandler) {
    return false;
  }

  dropRelationsForHandlerWithTag(handlerTag) {
    this.waitForRelations.delete(handlerTag);
    this.simultaneousRelations.delete(handlerTag);
  }

  reset() {
    this.waitForRelations.clear();
    this.simultaneousRelations.clear();
  }

}

exports.default = InteractionManager;
