// react-native-gesture-handler/lib/module/web/tools/GestureHandlerOrchestrator.js
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import { State } from '../../State';
import PointerTracker from './PointerTracker';
export default class GestureHandlerOrchestrator {
  // Private beacuse of Singleton
  // eslint-disable-next-line no-useless-constructor, @typescript-eslint/no-empty-function
  constructor() {
    _defineProperty(this, "gestureHandlers", []);

    _defineProperty(this, "awaitingHandlers", []);

    _defineProperty(this, "handlersToCancel", []);

    _defineProperty(this, "handlingChangeSemaphore", 0);

    _defineProperty(this, "activationIndex", 0);
  }

  scheduleFinishedHandlersCleanup() {
    if (this.handlingChangeSemaphore === 0) {
      this.cleanupFinishedHandlers();
    }
  }

  cleanHandler(handler) {
    handler.reset();
    handler.setActive(false);
    handler.setAwaiting(false);
    handler.setActivationIndex(Number.MAX_VALUE);
  }

  cleanupFinishedHandlers() {
    for (let i = this.gestureHandlers.length - 1; i >= 0; --i) {
      const handler = this.gestureHandlers[i];

      if (!handler) {
        continue;
      }

      if (this.isFinished(handler.getState()) && !handler.isAwaiting()) {
        this.gestureHandlers.splice(i, 1);
        this.cleanHandler(handler);
      }
    }
  }

  hasOtherHandlerToWaitFor(handler) {
    let hasToWait = false;
    this.gestureHandlers.forEach(otherHandler => {
      if (otherHandler && !this.isFinished(otherHandler.getState()) && this.shouldHandlerWaitForOther(handler, otherHandler)) {
        hasToWait = true;
        return;
      }
    });
    return hasToWait;
  }

  tryActivate(handler, event) {
    if (this.hasOtherHandlerToWaitFor(handler)) {
      this.addAwaitingHandler(handler);
    } else {
      this.makeActive(handler, event);
    }
  }

  cleanupAwaitingHandlers(handler) {
    for (let i = 0; i < this.awaitingHandlers.length; ++i) {
      if (!this.awaitingHandlers[i].isAwaiting() && this.shouldHandlerWaitForOther(this.awaitingHandlers[i], handler)) {
        this.cleanHandler(this.awaitingHandlers[i]);
        this.awaitingHandlers.splice(i, 1);
      }
    }
  }

  onHandlerStateChange(handler, newState, oldState, event) {
    this.handlingChangeSemaphore += 1;

    if (this.isFinished(newState)) {
      this.awaitingHandlers.forEach(otherHandler => {
        if (this.shouldHandlerWaitForOther(otherHandler, handler)) {
          if (newState === State.END) {
            otherHandler === null || otherHandler === void 0 ? void 0 : otherHandler.cancel(event);
            otherHandler === null || otherHandler === void 0 ? void 0 : otherHandler.setAwaiting(false);
          } else {
            this.tryActivate(otherHandler, event);
          }
        }
      });
    }

    if (newState === State.ACTIVE) {
      this.tryActivate(handler, event);
    } else if (oldState === State.ACTIVE || oldState === State.END) {
      if (handler.isActive()) {
        handler.sendEvent(event, newState, oldState);
      } else if (oldState === State.ACTIVE) {
        handler.sendEvent(event, newState, State.BEGAN);
      }
    } else if (oldState !== State.UNDETERMINED || newState !== State.CANCELLED) {
      handler.sendEvent(event, newState, oldState);
    }

    this.handlingChangeSemaphore -= 1;
    this.scheduleFinishedHandlersCleanup();

    if (this.awaitingHandlers.indexOf(handler) < 0) {
      this.cleanupAwaitingHandlers(handler);
    }
  }

  makeActive(handler, event) {
    const currentState = handler.getState();
    handler.setActive(true);
    handler.setShouldResetProgress(true);
    handler.setActivationIndex(this.activationIndex++);
    this.gestureHandlers.forEach(otherHandler => {
      // Order of arguments is correct - we check whether current handler should cancel existing handlers
      if (this.shouldHandlerBeCancelledBy(otherHandler, handler)) {
        this.handlersToCancel.push(otherHandler);
      }
    });

    for (let i = this.handlersToCancel.length - 1; i >= 0; --i) {
      var _this$handlersToCance;

      (_this$handlersToCance = this.handlersToCancel[i]) === null || _this$handlersToCance === void 0 ? void 0 : _this$handlersToCance.cancel(event);
    }

    this.awaitingHandlers.forEach(otherHandler => {
      if (this.shouldHandlerBeCancelledBy(otherHandler, handler)) {
        otherHandler === null || otherHandler === void 0 ? void 0 : otherHandler.cancel(event);
        otherHandler === null || otherHandler === void 0 ? void 0 : otherHandler.setAwaiting(true);
      }
    });
    handler.sendEvent(event, State.ACTIVE, State.BEGAN);

    if (currentState !== State.ACTIVE) {
      handler.sendEvent(event, State.END, State.ACTIVE);

      if (currentState !== State.END) {
        handler.sendEvent(event, State.UNDETERMINED, State.END);
      }
    }

    if (handler.isAwaiting()) {
      handler.setAwaiting(false);
      handler.end(event);
    }

    this.handlersToCancel = [];
  }

  addAwaitingHandler(handler) {
    let alreadyExists = false;
    this.awaitingHandlers.forEach(otherHandler => {
      if (otherHandler === handler) {
        alreadyExists = true;
        return;
      }
    });

    if (alreadyExists) {
      return;
    }

    this.awaitingHandlers.push(handler);
    handler.setAwaiting(true);
    handler.setActivationIndex(this.activationIndex++);
  }

  recordHandlerIfNotPresent(handler) {
    let alreadyExists = false;
    this.gestureHandlers.forEach(otherHandler => {
      if (otherHandler === handler) {
        alreadyExists = true;
        return;
      }
    });

    if (alreadyExists) {
      return;
    }

    this.gestureHandlers.push(handler);
    handler.setActive(false);
    handler.setAwaiting(false);
    handler.setActivationIndex(Number.MAX_SAFE_INTEGER);
  }

  shouldHandlerWaitForOther(handler, otherHandler) {
    return handler !== otherHandler && (handler.shouldWaitForHandlerFailure(otherHandler) || otherHandler.shouldRequireToWaitForFailure(handler));
  }

  canRunSimultaneously(gh1, gh2) {
    return gh1 === gh2 || gh1.shouldRecognizeSimultaneously(gh2) || gh2.shouldRecognizeSimultaneously(gh1);
  }

  shouldHandlerBeCancelledBy(handler, otherHandler) {
    const handlerPointers = handler.getTrackedPointersID();
    const otherPointers = otherHandler.getTrackedPointersID();

    if (!PointerTracker.shareCommonPointers(handlerPointers, otherPointers) && handler.getView() !== otherHandler.getView()) {
      return this.checkOverlap(handler, otherHandler);
    }

    if (this.canRunSimultaneously(handler, otherHandler)) {
      return false;
    }

    if (handler !== otherHandler && (handler.isAwaiting() || handler.getState() === State.ACTIVE)) {
      // For now it always returns false
      return handler.shouldBeCancelledByOther(otherHandler);
    }

    return true;
  }

  checkOverlap(handler, otherHandler) {
    // If handlers don't have common pointers, default return value is false.
    // However, if at least on pointer overlaps with both handlers, we return true
    // This solves issue in overlapping parents example
    // TODO: Find better way to handle that issue, for example by activation order and handler cancelling
    const handlerPointers = handler.getTrackedPointersID();
    const otherPointers = otherHandler.getTrackedPointersID();
    let overlap = false;
    handlerPointers.forEach(pointer => {
      const handlerX = handler.getTracker().getLastX(pointer);
      const handlerY = handler.getTracker().getLastY(pointer);

      if (handler.getEventManager().isPointerInBounds({
        x: handlerX,
        y: handlerY
      }) && otherHandler.getEventManager().isPointerInBounds({
        x: handlerX,
        y: handlerY
      })) {
        overlap = true;
      }
    });
    otherPointers.forEach(pointer => {
      const otherX = otherHandler.getTracker().getLastX(pointer);
      const otherY = otherHandler.getTracker().getLastY(pointer);

      if (handler.getEventManager().isPointerInBounds({
        x: otherX,
        y: otherY
      }) && otherHandler.getEventManager().isPointerInBounds({
        x: otherX,
        y: otherY
      })) {
        overlap = true;
      }
    });
    return overlap;
  }

  isFinished(state) {
    return state === State.END || state === State.FAILED || state === State.CANCELLED;
  }

  static getInstance() {
    if (!GestureHandlerOrchestrator.instance) GestureHandlerOrchestrator.instance = new GestureHandlerOrchestrator();
    return GestureHandlerOrchestrator.instance;
  }

}

_defineProperty(GestureHandlerOrchestrator, "instance", void 0);
