// react-native-gesture-handler/lib/module/web/handlers/PinchGestureHandler.js
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import { State } from '../../State';
import { DEFAULT_TOUCH_SLOP } from '../constants';
import { EventTypes } from '../interfaces';
import GestureHandler from './GestureHandler';
import ScaleGestureDetector from '../detectors/ScaleGestureDetector';
export default class PinchGestureHandler extends GestureHandler {
  constructor(...args) {
    super(...args);

    _defineProperty(this, "scale", 1);

    _defineProperty(this, "velocity", 0);

    _defineProperty(this, "startingSpan", 0);

    _defineProperty(this, "spanSlop", 0);

    _defineProperty(this, "scaleDetectorListener", {
      onScaleBegin: detector => {
        this.startingSpan = detector.getCurrentSpan();
        return true;
      },
      onScale: (detector, event) => {
        const prevScaleFactor = this.scale;
        this.scale *= detector.getScaleFactor(this.tracker.getTrackedPointersCount());
        const delta = detector.getTimeDelta();

        if (delta > 0) {
          this.velocity = (this.scale - prevScaleFactor) / delta;
        }

        if (Math.abs(this.startingSpan - detector.getCurrentSpan()) >= this.spanSlop && this.currentState === State.BEGAN) {
          this.activate(event);
        }

        return true;
      },
      onScaleEnd: (_detector, _event) => {}
    });

    _defineProperty(this, "scaleGestureDetector", new ScaleGestureDetector(this.scaleDetectorListener));
  }

  init(ref, propsRef) {
    super.init(ref, propsRef);
    this.setShouldCancelWhenOutside(false);
    this.spanSlop = DEFAULT_TOUCH_SLOP;
  }

  updateGestureConfig({
    enabled = true,
    ...props
  }) {
    super.updateGestureConfig({
      enabled: enabled,
      ...props
    });
    this.enabled = enabled;
  }

  transformNativeEvent(_event) {
    return {
      focalX: this.scaleGestureDetector.getFocusX(),
      focalY: this.scaleGestureDetector.getFocusY(),
      velocity: this.velocity,
      scale: this.scale
    };
  }

  onPointerDown(event) {
    super.onPointerDown(event);
    this.tracker.addToTracker(event);

    if (this.tracker.getTrackedPointersCount() < 2) {
      return;
    }

    if (this.tracker.getTrackedPointersCount() > 1) {
      this.tryBegin(event);
      this.scaleGestureDetector.onTouchEvent(event, this.tracker);
    }
  }

  onPointerUp(event) {
    if (this.tracker.getTrackedPointersCount() > 1) {
      this.scaleGestureDetector.onTouchEvent(event, this.tracker);
      this.tracker.removeFromTracker(event.pointerId);
    } else {
      this.tracker.removeFromTracker(event.pointerId);

      if (this.currentState !== State.ACTIVE) {
        return;
      }

      this.scaleGestureDetector.onTouchEvent(event, this.tracker);
    }

    if (this.currentState === State.ACTIVE && this.tracker.getTrackedPointersCount() < 2) {
      this.end(event);
    } else if (event.eventType === EventTypes.UP && this.currentState !== State.BEGAN) {
      this.fail(event);
    }
  }

  onPointerMove(event) {
    if (this.tracker.getTrackedPointersCount() < 2) {
      return;
    }

    this.tracker.track(event);
    this.scaleGestureDetector.onTouchEvent(event, this.tracker);
    super.onPointerMove(event);
  }

  onPointerOutOfBounds(_event) {//
  }

  onPointerCancel(_event) {
    this.reset();
  }

  tryBegin(event) {
    if (this.currentState !== State.UNDETERMINED) {
      return;
    }

    this.resetProgress();
    this.begin(event);
  }

  activate(event, force) {
    if (this.currentState !== State.ACTIVE) {
      this.resetProgress();
    }

    super.activate(event, force);
  }

  onReset() {
    this.resetProgress();
  }

  resetProgress() {
    if (this.currentState === State.ACTIVE) {
      return;
    }

    this.velocity = 0;
    this.scale = 1;
  }

}
