// react-native-gesture-handler/lib/module/web/handlers/TapGestureHandler.js
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import { State } from '../../State';
import { EventTypes } from '../interfaces';
import GestureHandler from './GestureHandler';
const DEFAULT_MAX_DURATION_MS = 500;
const DEFAULT_MAX_DELAY_MS = 200;
const DEFAULT_NUMBER_OF_TAPS = 1;
const DEFAULT_MIN_NUMBER_OF_POINTERS = 1;
export default class TapGestureHandler extends GestureHandler {
  constructor(...args) {
    super(...args);

    _defineProperty(this, "maxDeltaX", Number.MIN_SAFE_INTEGER);

    _defineProperty(this, "maxDeltaY", Number.MIN_SAFE_INTEGER);

    _defineProperty(this, "maxDistSq", Number.MIN_SAFE_INTEGER);

    _defineProperty(this, "maxDurationMs", DEFAULT_MAX_DURATION_MS);

    _defineProperty(this, "maxDelayMs", DEFAULT_MAX_DELAY_MS);

    _defineProperty(this, "numberOfTaps", DEFAULT_NUMBER_OF_TAPS);

    _defineProperty(this, "minNumberOfPointers", DEFAULT_MIN_NUMBER_OF_POINTERS);

    _defineProperty(this, "currentMaxNumberOfPointers", 1);

    _defineProperty(this, "startX", 0);

    _defineProperty(this, "startY", 0);

    _defineProperty(this, "offsetX", 0);

    _defineProperty(this, "offsetY", 0);

    _defineProperty(this, "lastX", 0);

    _defineProperty(this, "lastY", 0);

    _defineProperty(this, "waitTimeout", void 0);

    _defineProperty(this, "delayTimeout", void 0);

    _defineProperty(this, "tapsSoFar", 0);
  }

  init(ref, propsRef) {
    super.init(ref, propsRef);
    this.setShouldCancelWhenOutside(true);
  }

  updateGestureConfig({
    enabled = true,
    ...props
  }) {
    super.updateGestureConfig({
      enabled: enabled,
      ...props
    });
    this.enabled = enabled;

    if (this.config.numberOfTaps !== undefined) {
      this.numberOfTaps = this.config.numberOfTaps;
    }

    if (this.config.maxDurationMs !== undefined) {
      this.maxDurationMs = this.config.maxDurationMs;
    }

    if (this.config.maxDelayMs !== undefined) {
      this.maxDelayMs = this.config.maxDelayMs;
    }

    if (this.config.maxDeltaX !== undefined) {
      this.maxDeltaX = this.config.maxDeltaX;
    }

    if (this.config.maxDeltaY !== undefined) {
      this.maxDeltaY = this.config.maxDeltaY;
    }

    if (this.config.maxDistSq !== undefined) {
      this.maxDistSq = this.config.maxDistSq;
    }

    if (this.config.minPointers !== undefined) {
      this.minNumberOfPointers = this.config.minPointers;
    }
  }

  resetConfig() {
    super.resetConfig();
    this.maxDeltaX = Number.MIN_SAFE_INTEGER;
    this.maxDeltaY = Number.MIN_SAFE_INTEGER;
    this.maxDistSq = Number.MIN_SAFE_INTEGER;
    this.maxDurationMs = DEFAULT_MAX_DURATION_MS;
    this.maxDelayMs = DEFAULT_MAX_DELAY_MS;
    this.numberOfTaps = DEFAULT_NUMBER_OF_TAPS;
    this.minNumberOfPointers = DEFAULT_MIN_NUMBER_OF_POINTERS;
  }

  transformNativeEvent(event) {
    return {
      x: event.offsetX,
      y: event.offsetY,
      absoluteX: event.x,
      absoluteY: event.y
    };
  }

  clearTimeouts() {
    clearTimeout(this.waitTimeout);
    clearTimeout(this.delayTimeout);
  }

  startTap(event) {
    this.clearTimeouts();
    this.waitTimeout = setTimeout(() => this.fail(event), this.maxDurationMs);
  }

  endTap(event) {
    this.clearTimeouts();

    if (++this.tapsSoFar === this.numberOfTaps && this.currentMaxNumberOfPointers >= this.minNumberOfPointers) {
      this.activate(event);
    } else {
      this.delayTimeout = setTimeout(() => this.fail(event), this.maxDelayMs);
    }
  } //Handling Events


  onPointerDown(event) {
    super.onPointerDown(event);
    this.tracker.addToTracker(event);
    this.trySettingPosition(event);

    if (this.tracker.getTrackedPointersCount() > 1) {
      this.onPointerAdd(event);
    } else {
      this.lastX = this.tracker.getLastAvgX();
      this.lastY = this.tracker.getLastAvgY();
    }

    this.updateState(event);
  }

  onPointerAdd(_event) {
    this.offsetX += this.lastX - this.startX;
    this.offsetY += this.lastY = this.startY;
    this.lastX = this.tracker.getLastAvgX();
    this.lastY = this.tracker.getLastAvgY();
    this.startX = this.lastX;
    this.startY = this.lastY;
  }

  onPointerUp(event) {
    if (this.tracker.getTrackedPointersCount() > 1) {
      this.tracker.removeFromTracker(event.pointerId);
      this.onPointerRemove(event);
    } else {
      this.lastX = this.tracker.getLastAvgX();
      this.lastY = this.tracker.getLastAvgY();
      this.tracker.removeFromTracker(event.pointerId);
    }

    this.updateState(event);
  }

  onPointerRemove(_event) {
    this.offsetX += this.lastX - this.startX;
    this.offsetY += this.lastY = this.startY;
    this.lastX = this.tracker.getLastAvgX();
    this.lastY = this.tracker.getLastAvgY();
    this.startX = this.lastX;
    this.startY = this.lastY;
  }

  onPointerMove(event) {
    this.trySettingPosition(event);
    this.lastX = this.tracker.getLastAvgX();
    this.lastY = this.tracker.getLastAvgY();
    this.updateState(event);
  }

  onPointerOutOfBounds(event) {
    this.trySettingPosition(event);
    this.lastX = this.tracker.getLastAvgX();
    this.lastY = this.tracker.getLastAvgY();
    this.updateState(event);
  }

  onPointerCancel(event) {
    this.tracker.resetTracker();
    this.fail(event);
  }

  updateState(event) {
    if (this.currentMaxNumberOfPointers < this.tracker.getTrackedPointersCount()) {
      this.currentMaxNumberOfPointers = this.tracker.getTrackedPointersCount();
    }

    if (this.shouldFail()) {
      this.fail(event);
      return;
    }

    switch (this.currentState) {
      case State.UNDETERMINED:
        if (event.eventType === EventTypes.DOWN) {
          this.begin(event);
        }

        this.startTap(event);
        break;

      case State.BEGAN:
        if (event.eventType === EventTypes.UP) {
          this.endTap(event);
        }

        if (event.eventType === EventTypes.DOWN) {
          this.startTap(event);
        }

        break;

      default:
        break;
    }
  }

  trySettingPosition(event) {
    if (this.currentState !== State.UNDETERMINED) {
      return;
    }

    this.offsetX = 0;
    this.offsetY = 0;
    this.startX = event.x;
    this.startY = event.y;
  }

  shouldFail() {
    const dx = this.lastX - this.startX + this.offsetX;

    if (this.maxDeltaX !== Number.MIN_SAFE_INTEGER && Math.abs(dx) > this.maxDeltaX) {
      return true;
    }

    const dy = this.lastY - this.startY + this.offsetY;

    if (this.maxDeltaY !== Number.MIN_SAFE_INTEGER && Math.abs(dy) > this.maxDeltaY) {
      return true;
    }

    const dist = dy * dy + dx * dx;
    return this.maxDistSq !== Number.MIN_SAFE_INTEGER && dist > this.maxDistSq;
  }

  activate(event) {
    super.activate(event);

    if (!this.isAwaiting()) {
      this.end(event);
    }
  }

  onCancel() {
    this.resetProgress();
    this.clearTimeouts();
  }

  resetProgress() {
    this.tapsSoFar = 0;
    this.currentMaxNumberOfPointers = 0;
  }

}
